src/darwin/Framework/CHIP/ServerEndpoint/MTRServerAttribute.mm

```diff
@@ -0,0 +1,210 @@
+        return YES;
+    }
+
+    MTRDeviceController * deviceController = _deviceController;
+    if (deviceController == nil || !deviceController.running) {
+        // Controller went away.  Make sure we are invalidated.
+        [self invalidate];
+    }
+
+    if (_state == MTRServerEndpointStateDefunct) {
+        MTR_LOG_ERROR("Cannot %@ on attribute %llx on no-longer-active endpoint", action, _attributeID.unsignedLongLongValue);
+        return NO;
+    }
+
+    return YES;
+}
+
+- (BOOL)setValue:(NSDictionary<NSString *, id> *)value
+{
+    if (![self ensureState:@"change value"]) {
+        return NO;
+    }
+
+    id serializedValue;
+    id dataType = value[MTRTypeKey];
+    if ([MTRArrayValueType isEqual:dataType]) {
+        id dataValue = value[MTRValueKey];
+        if (![dataValue isKindOfClass:NSArray.class]) {
+            MTR_LOG_ERROR("MTRServerAttribute value claims to be a list but isn't: %@", value);
+            return NO;
+        }
+        NSArray * dataValueList = dataValue;
+        auto * listValue = [NSMutableArray arrayWithCapacity:dataValueList.count];
+        if (listValue == nil) {
+            return NO;
+        }
+        for (id item in dataValueList) {
+            NSError * encodingError;
+            NSData * encodedItem = MTREncodeTLVFromDataValueDictionary(item, &encodingError);
+            if (encodedItem == nil) {
+                return NO;
+            }
+            [listValue addObject:encodedItem];
+        }
+        serializedValue = listValue;
+    } else {
+        NSError * encodingError;
+        serializedValue = MTREncodeTLVFromDataValueDictionary(value, &encodingError);
+        if (serializedValue == nil) {
+            return NO;
+        }
+    }
+
+    // We serialized properly, so should be good to go on the value.
+    _value = [value copy];
+
+    if (_state == MTRServerEndpointStateInitializing) {
+        _serializedValue = serializedValue;
+    } else {
+        // Must be non-nil if ensureState already happened.
+        MTRDeviceController * deviceController = _deviceController;
+        [deviceController asyncDispatchToMatterQueue:^{
+            auto changed = ![self->_serializedValue isEqual:serializedValue];
+            self->_serializedValue = serializedValue;
+            if (changed) {
+                MatterReportingAttributeChangeCallback(self->_parentCluster.mEndpointId, self->_parentCluster.mClusterId, static_cast<AttributeId>(self->_attributeID.unsignedLongLongValue));
+            }
+        }
+                                        errorHandler:nil];
+    }
+
+    return YES;
+}
+
+- (BOOL)associateWithController:(MTRDeviceController *)controller
+{
+    if (_state != MTRServerEndpointStateInitializing) {
+        MTR_LOG_ERROR("Cannot associate MTRServerAttribute in state %lu with controller", static_cast<unsigned long>(_state));
+        return NO;
+    }
+
+    _state = MTRServerEndpointStateActive;
+    _deviceController = controller;
```

Comment:
This object is not thread safe. This seems desirable, since we have multiple users of it (the client and device-controller).

Reply:
Hmmm.  So in terms of the threading behavior, the way this will be used is:

1. Client sets up the object, while _state is Initializing.
2. Client hands object to device controller.  That 
    1. Immediately sets state to Active (before returning).
    2. Sets some other members.
    3. Queues a task on the Matter queue to do other things with the object.

The "other members" do get accessed on the Matter queue, so we might need some synchronization there to ensure that the write to them is actually complete on this thread before the task runs.  But I would think that dispatch_async acts as a memory barrier in general; otherwise all sorts of problems would arise with memory writes "before" the dispatch not being done before the thing being dispatched runs, no?

Past that, is there a specific concern about threading here?

Reply:
Per Slack discussion: we might want to have multiple different API client threads touching the same endpoint/cluster/attribute objects, in this setup with mutable access grants and mutable attribute state, and we need to make that safe.  I'll do a followup to add locks to handle that.

Reply:
https://github.com/project-chip/connectedhomeip/issues/31756